### PR TITLE
JS: improve performance by combining stages

### PR DESCRIPTION
**Pull request by erik-krogh**
TL;DR: [~~7%~~ ~9% performance gain on the security suite](https://github.com/dsp-testing/erik-krogh-dca/tree/run/staging-even-more-default-security/reports) by grouping stages. 

---- 

Some of the performance work I've done recently has included adding more `cached` predicates.  
Adding `cached` predicates has improved performance, but it has also introduced more stages in the computation (since a stage is a strongly connected component of `cached` predicates). 

Having more stages introduces redundant computations, as non-cached predicates are recomputed in each stage. 
One solution would be to add more `cached` predicates, which would add even more stages, which would again introduce redundant computations. So that is not a good option. 

So I tried to remove all `cached` annotations to see what would happen (somehow that didn't hit the BDD limit). 
The result was [38% faster single-query performance (on `TaintedPath.ql`)](https://github.com/dsp-testing/erik-krogh-dca/tree/run/nocache-default-taintedpath-2/reports).   
Removing all `cached` annotations will destroy multi-query performance, so that is not a good option.  

What I've done instead is to reduce the number of stages computed at runtime, by grouping stages into extended stages by adding dependencies between the `cached` predicates that were previously in different stages. 
You can read the qldoc in [ExtendedStaging.qll](https://github.com/github/codeql/blob/7066fb3e2668fa17087df449f421f23fc15f5b6a/javascript/ql/src/semmle/javascript/ExtendedStaging.qll) for more details on how this is implemented. 

I decided which stages to group together by looking at which `cached` predicates got evaluated by a taint-tracking query. 
I've been somewhat aggressive in grouping different stages together, as I found a strong correlation between less stages and more performance. 
([An earlier less aggressive attempt gave a 2.5% performance gain](https://github.com/dsp-testing/erik-krogh-dca/tree/run/staging-default-security/reports)). 

This means that a query evaluation will sometimes compute `cached` predicates that are not needed for that query.  

[The resulting performance gain is ~9% on the security suite](https://github.com/dsp-testing/erik-krogh-dca/tree/run/staging-even-more-default-security/reports). 

--- 

I would have liked to group even more stages together, but some groupings are impossible with the current approach. 
Grouping more stages together would require language support.  
Maybe something like: 
```CodeQL
cached[ast]
predicate myAstPredicate() { ... }
```

### CODE REVIEW

**Comment by aschackmull** [comment]
> Grouping more stages together would require language support.

There is already some language support for grouping stages: If you have several `cached` predicates in a module you can mark the module as `cached` - that will group all the contained `cached` predicates into a single stage.

**Comment by erik-krogh** [comment]
> There is already some language support for grouping stages: If you have several `cached` predicates in a module you can mark the module as `cached` - that will group all the contained `cached` predicates into a single stage.

Thanks. I didn't realize that.  

But these `cached` predicates generally belong where they are, in different files and classes.   
It would require a lot of refactorization to get the them into a single `cached` module.

**Comment by aschackmull** [comment]
Note that you can use cached modules to introduce stage-level dependencies that aren't predicate-level dependencies, which can be convenient to avoid spurious recursion when merging multiple stages into one. See also the C# code (`csharp/ql/src/semmle/code/csharp/Caching.qll`), which does a lot of these kinds of tricks.

**Comment by erik-krogh** [comment]
> Note that you can use cached modules to introduce stage-level dependencies that aren't predicate-level dependencies, which can be convenient to avoid spurious recursion when merging multiple stages into one. 

Nice. I used that to get one more `cached` predicate into the `ast` stage. (Currently evaluating that). 

I see the convenience of using `cached` modules. 
But on the other hand, the predicates in those modules gets evaluated by the runtime.  
With my approach, everything is eliminated by optimizer.

**Comment by asgerf** [comment]
> But on the other hand, the predicates in those modules gets evaluated by the runtime

I don't follow. I don't think @aschackmull was suggesting that you should put these phantom dependency-predicates into a cached module.

Instead, could you try putting the predicates themselves into a cached module, and remove the phantom dependencies? The original predicates can then be replaced with predicate aliases or member predicates with a `pragma[inline]`.

**Comment by erik-krogh** [comment]
> could you try putting the predicates themselves into a cached module, and remove the phantom dependencies? The original predicates can then be replaced with predicate aliases or member predicates with a pragma[inline].

I could, but I prefer not to. 
Some of the predicates depend on other private predicates in the class/file/module, so moving the predicates into a single cached module would require significant refactorization.  
And the predicates don't naturally belong in one file. E.g. the extended `dataflow` stage span 7 files. 

Right now I'm leaning towards making cached modules in the same style as the [C# implementation](https://github.com/github/codeql/blob/main/csharp/ql/src/semmle/code/csharp/Caching.qll).   
Because that approach always works, even if there is a negative dependency between the predicates I want to have in the same stage.

**Comment by asgerf** [comment]
Hm, yeah I can see it's nice to be able to keep predicates in their original places like that. I must admit I was a little sceptical at first, but I'm slowing coming around to this way of doing it.

> Right now I'm leaning towards making cached modules in the same style as the C# implementation.

Agreed. Do you want to go ahead and migrate to this style? Hopefully it won't upset any join orders.

**Comment by erik-krogh** [comment]
> Agreed. Do you want to go ahead and migrate to this style? Hopefully it won't upset any join orders.

I already checked that in an evaluation. Performance is within the margin of error.

**Comment by erik-krogh** [comment]
> Do you want to go ahead and migrate to this style?

Done.

File: javascript/ql/src/semmle/javascript/GlobalAccessPaths.qll [hunk 1]
```diff
@@ -440,6 +439,7 @@ module AccessPath {
      *
      * Only has a result if there exists both a read and write of the access-path within `bb`.
      */
+    pragma[nomagic]
```
Comment (erik-krogh) [inline javascript/ql/src/semmle/javascript/GlobalAccessPaths.qll:442]: This prevented a catastrophic join-order when I did the "remove all `cached` annotations"-experiment.  
I don't know if it has an effect with the current implementation, but I'm keeping it anyway.

File: javascript/ql/src/semmle/javascript/ExtendedStaging.qll [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+   *   AST::ASTNode::isAmbientInternal
+   */
+  cached
+  module Ast {
+    /**
+     * Always holds.
+     * Ensures that a predicate is evaluated as part of the Ast stage.
+     */
+    cached
+    predicate ref() { 1 = 1 }
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/ExtendedStaging.qll:70]: No, the behavior would be the same if I used `any()` instead. 
The optimizer compiles it down to a constant in both cases.   

I just preferred `1 = 1`, partially because the syntax highlighting sticks out more. 
(If I used `any()`, then the `any()` would look similar to the other references in the `backref` predicate).

File: javascript/ql/src/javascript.qll [hunk 3]
```diff
@@ -34,6 +33,7 @@ import semmle.javascript.GlobalAccessPaths
 import semmle.javascript.HTML
 import semmle.javascript.HtmlSanitizers
 import semmle.javascript.InclusionTests
+import semmle.javascript.internal.CachedStages
```
Comment (asgerf) [inline javascript/ql/src/javascript.qll:36]: The point of being `internal` is sort of lost if we import it from `javascript.qll`.

Instead add private imports from the files that need it.

**Review by asgerf — APPROVED** [review]
🚀
